driver/src/orderbook/streamed/balance.rs

```diff
@@ -2,63 +2,104 @@ use super::*;
 use error::Error;
 use serde::{Deserialize, Serialize};
 
+/// The balance of a token for a user.
 #[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
 pub struct Balance {
     balance: U256,
-    pub pending_deposit: Option<Flux>,
-    pub pending_withdraw: Option<Flux>,
+    deposit: Flux,
+    withdraw: Flux,
 }
 
 impl Balance {
-    pub fn deposit(&mut self, new_deposit: Flux, current_batch_id: BatchId) {
-        self.update_deposit_balance(current_batch_id);
-        match self.pending_deposit.as_mut() {
-            Some(deposit) => {
-                deposit.amount += new_deposit.amount;
-                deposit.batch_id = new_deposit.batch_id
-            }
-            None => self.pending_deposit = Some(new_deposit),
-        }
+    pub fn deposit(
+        &mut self,
+        amount: U256,
+        deposit_batch_id: BatchId,
+        current_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        self.apply_existing_deposit(current_batch_id)?;
+        // Works like in the smart contract: If there is an existing deposit we override the
+        // batch id and add to the amount. If there is no existing deposit then amount is already 0.
+        let deposit_amount_including_this_deposit = self
+            .deposit
+            .amount
+            .checked_add(amount)
+            .ok_or(Error::MathOverflow)?;
+        self.deposit.batch_id = deposit_batch_id;
+        self.deposit.amount = deposit_amount_including_this_deposit;
+        Ok(())
+    }
+
+    pub fn withdraw_request(&mut self, amount: U256, batch_id: BatchId) {
+        self.withdraw.batch_id = batch_id;
+        self.withdraw.amount = amount;
     }
 
-    pub fn withdraw(&mut self, amount: U256, current_batch_id: BatchId) -> Result<(), Error> {
-        self.update_deposit_balance(current_batch_id);
-        self.pending_withdraw = None;
-        self.balance = self
-            .balance
+    pub fn withdraw(&mut self, amount: U256, batch_id: BatchId) -> Result<(), Error> {
+        if self.withdraw.batch_id >= batch_id {
+            return Err(Error::WithdrawEarlierThanRequested(self.withdraw.batch_id));
+        }
+        if self.withdraw.amount < amount {
+            return Err(Error::WithdrawMoreThanRequested(self.withdraw.amount));
+        }
+        let balance_excluding_withdraw = self.get_balance_internal(batch_id, false)?;
+        let new_balance = balance_excluding_withdraw
             .checked_sub(amount)
-            .ok_or(Error::MathUnderflow)?;
+            .ok_or(Error::WithdrawMoreThanBalance(balance_excluding_withdraw))?;
+        self.balance = new_balance;
+        // If there is a valid deposit then new_balance already includes it so we need to clear it
+        // now.
+        self.clear_pending_deposit_if_possible(batch_id);
```

Comment:
@fleupold 
> The thing I dislike about this method is that it can only be safely called if before we have called get_balance_internal(batch_id, false) and re-write the result back to balance afterwards (otherwise we lose the pending deposit and and solution). This expectation is very implicit.

Because it is called "clear" and the implementation is simple I disagree that this is too implicit.

> Would it be possible to rename get_balance_internal into updated_and_get_balance and do the persistence as well as the clearing directly in there?

get_balance should not be mutable but I will add another function that does this.

Reply:
I couldn't' do this in a way that I find to improve the code so I am keeping it like this. Compared to when you made your review there is a new comment in the code explaining why we are clearing the pending deposit.

Reply:
Just curious, is it possible to `apply_deposit_if_possilbe` and then apply the withdrawal separately? The reason I suggest this is to reduce the number of places that mutating the deposit flux can happen.

Reply:
The problem is that I want to guarantee that we don't modify anything if an error happens. We can error if the withdraw amount is larger than `balance_excluding_withdraw` in the `checked_sub`. If I had applied the deposit then the deposit would stay applied even though there was an error.

Reply:
But if we error doesn't that likely mean that this current `Balance` state will be in an incorrect state anyway?

Reply:
It means that the event is not consistent with the state but the state is still consistent with the pats events it got.
From the perspective of just `State` doing it this way makes more sense. If you do not yet know how future code will use it then it makes sense that you don't invalidate the state on error because that code might want to reuse the state and try a different event. If we invalidated state on error this wouldn't be possible.
Even with specific knowledge of how state is probably going to be used in our code with the snapshot we might still want this. Like let's say we get a new event and it fails to apply to state and we decide to "blacklist" this event and keep going. Without the guarantee we would have to reconstruct from the snapshot because state got invalidated. With the guarantee we can keep using state for the next event.

Reply:
> It means that the event is not consistent with the state but the state is still consistent with the past events it got.

This is not true though, because the `State` will not be valid for that batch anymore since it is ignoring events. So it should arguably not be considered at all anymore for any batch after the last successful snapshot it was included in.

> might want to reuse the state and try a different event.

In that case not modifying the state does make sense, but I think it makes sense to do it atomically by doing a `Balances::clone()` and then applying the event, and then replacing the balances in case it succeeds, as it would make the code easier to follow and have the state have to think less about side-effects of functions and in what order those side-effects get applied.

> With the guarantee we can keep using state for the next event.

But that would mean that ultimately the state at the end of the batch just ignored an event and is invalid anyway, and using that invalid state and passing it to the solver may lead to invalid solutions which is worse than just not including it, which could only lead to suboptimal solutions.

Reply:
I feel like this is another situation where the state would benefit from getting all of the events for a batch at once so it can apply things in the order that is easiest for it.

Reply:
> In that case not modifying the state does make sense, but I think it makes sense to do it atomically by doing a Balances::clone() 

True. By not holding up the guarantee in Balance but in State we can simplify the code in Balance. It still feels better to me to do this in Balance too but I agree it would be simpler outside of it. If Balance contained heap data I would be more opposed to it but maybe I am too picky here. On the other hand, I don't think this function is that complicated either. We could take this point to the extreme and say we don't need to guarantee this in State either because we can clone State before applying an event. In that case clearly it would be too expensive - but for Balance not.

Reply:
Its copying a stack copy of less than 100 bytes, so basically free :wink:.

>  On the other hand, I don't think this function is that complicated either

you're right it isn't but I still strongly believe there is no point holding a `Balance` struct around once it errors - If any event is not applied to it then it is wrong and can't be included in the orderbook so I think we should just modify and if there is an error black list it.

The only way this extra complexity will be worth it is if the events are being applied in the wrong order, and then the events need to tentatively be applied before. In that case you will need to keep a queue of events that need to be re-applied for that specific balance in order for it to "catch back up". I think this will add a lot of extra complexity that is not necessarily useful yet.